### ISSUE TITLE
Web based version

### ISSUE THREAD

**Issue by okohll** [biolab/orange3#1419]
Hi, not sure this is really an issue so much as an 'epic' but are there any plans for a web based version that one could host sometime in the future? A big job I know but that would be awesome.

**Comment by ajdapretnar** [comment]
@okohll This has been on our wish list for a long time, but as you said, it's a big job. Currently our main priority is to assure stability of Orange 3 and introduce industry standards where possible (GSoC's to pandas dataframe project).
It would be **really** cool to have this and yes, perhaps sometime in the future, our common dreams will come true. :stars:

**Comment by brylie** [comment]
What are some of the tasks involved in the 'web port'? I am a full-time web application developer, so this would be within my comfort zone. 

May we please break the 'web port' idea into some practical tasks, and track the tasks via an epic label?

**Comment by kernc** [comment]
1. Design a core **pure JS API** and protocols for interaction between (2), (3) and (4), so the whole still works together if any of  (2), (3), (4) or (5) is switched for something else in the future.
2. Re-architecture Orange into a **client-server** application with the **backend** Python server handling user's JS events. The backend architecture should support independent and parallel (multi-core) widget execution, so the widgets should run in different Python processes and at the same time, preferably, have an option to share the same copy of data (unless they declare otherwise) — kind of like what [joblib](https://pythonhosted.org/joblib/) does. For convenience, any solution should possibly depend on [ipyparallel](https://ipyparallel.readthedocs.io/en/latest/). Hopefully, they have figured the hard stuff out.
3. Re-engineer the **widget architecture** which is currently tailored to GUI programming using Qt. Perhaps try [Bokeh](http://bokeh.pydata.org/en/latest/) and see how extensible it is in terms of different/custom plot types. [Bokeh interactive widgets example](https://demo.bokehplots.com/apps/movies) (and its [code](https://github.com/bokeh/bokeh/blob/master/examples/app/movies/main.py)).
4. Figure out how to render various visualizations on the web (using [D3.js](https://d3js.org/) or something higher-/lower-level) with possibility of [various libs](https://syntagmatic.github.io/parallel-coordinates) for **various plot types**. For maximum existing JS plotting ecosystem reuse.
5. **Web app** UI, canvas implemented with modern (but mature and stable!) JS/CSS libs as a reactive web app -- this is the easiest part that should, according to proper risk management, come **late**.
6. Port existing widgets.

All above but the last are largely architectural, non-easy tasks. Should be tracked as individual issues as they become worked on. Epic label sounds quite appropriate.

**Comment by brylie** [comment]
May we please open tasks fore the above?

**Comment by kernc** [comment]
Will anyone be working on them?

**Comment by brylie** [comment]
@kernc the tasks can be in an 'icebox' status until there are development resources available. This is a common practice in agile projects.

**Comment by ajdapretnar** [comment]
@brylie I think on Github it should just work with tags, no? @kernc How about a checklist? My concern here is that it's such an abstract checklist one can easily lose track of actual tasks. So either we leave it at that or create a more detailed checklist with specific tasks to be address (could be even solved with several PRs).
For now, I'll just add 'enhancement' tag to the issue.

**Comment by brylie** [comment]
@ajdapretnar by 'icebox status' I meant tag. E.g. when using [Waffle.io](https://waffle.io/), each Kanban column is simply assigned a Github tag.

**Comment by kernc** [comment]
Yes, but this is not Waffle, rather an issue tracker. While it does make sense to create subtasks in order to discuss the associated details, the consensus of the maintainers unfortunately was not to make such an endeavor as to port Orange to the web any time soon. Thus, above task scheduling approach makes little sense.

If you feel like designing the questions and answers to any particular relevant subtask, you are welcome to open it yourself. In other words, in your opinion, what is the likelihood of any such subtask to bear status "in progress" or "done" in the next year or so?

**Comment by brylie** [comment]
Kanban is also an issue tracker, just with more statuses than a common 'todo list'.

According to the proposed list above, we could begin with an API design. The API design could inform subsequent frontend and backend development efforts.

**Comment by kernc** [comment]
When you say "we could begin," who do you mean by we? Will you be leading the effort?

**Comment by brylie** [comment]
Good question. I will have some availability around October.

**Comment by brylie** [comment]
@kernc and @ajdapretnar, another developer (@marla-singer] and I are now sketching out this idea.

**Comment by brylie** [comment]
We produced a quick video update describing our idea and project plans.

[![apinf-experimentationmonday-mlrestapi-status1](https://cloud.githubusercontent.com/assets/17307/24863966/7db3f896-1e0b-11e7-80aa-e5966b59eee6.jpg)](https://youtu.be/nOwfYGaCr3w)

We welcome any ideas or feedback related to our presentation.

**Comment by kernc** [comment]
Went over the video and your write-up. Am watching the repo. At this early vague stage I have few constructive comments. Perhaps open issues for subtasks to be discussed in. Glad to have you here. :+1:

**Comment by brylie** [comment]
Thanks @kernc 

While doing research and preplanning, I came across a specification called [Protocols and Structures for Inference Machine Learning as a Service](http://psikit.net/) (PsiKit), which even has [reverence implementation client/server code](https://github.com/psi-project) (from 2013).

What are your thoughts on using PsiKit specification as the basis of the Orange3 ML REST API?

**Comment by kernc** [comment]
Without investing significant time in the nitty gritty schema details, I find it overall pretty reasonable; perhaps lacking a few features. I had the following ideas come to mind:
1. Transformers need to distinct between _fit_ and _transform_ steps. Using [StandardScaler](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html#sklearn.preprocessing.StandardScaler) as an example, it first needs to learn the mean and deviation on the training data, transform the training data, but also transform any subsequent test/additional data with the same mean/deviation parameters. In this way, it acts more like a learner, but it sounds like a transformer in the sense of its [definition](http://psikit.net/spec/#transformers).
2. Interactive (or otherwise) visualization does not seem to be covered. I guess visualization end-points act as transformers, outputting selected or annotated subsets.
3. How would I define e.g. a [Test&Score](http://docs.orange.biolab.si/3/visual-programming/widgets/evaluation/testlearners.html) or a [parameter tuning](http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html#sklearn.model_selection.GridSearchCV) end-point? It accepts, among other things, a _learner_ (or an array thereof; that, e.g., accepts a nominal target) and any optional parameters, and it outputs an array of scores, or an array of best parameterized learners. Is this a learner or a transformer? Can anything this complex be defined with current PSI schema? 
4. `resource-list` objects would want to be queried, filtered on. E.g. give me just the data relations that have size > 1000. Or, give me the trained predictors that were trained in such a way that they accept data as `[int, int, int]` and output a particular enum ...
5. When [using a predictor](http://psikit.net/spec/#using-a-predictor), contrary to the examples, the verb should be POST. It should always be POST for complex, resource-intensive queries.
6. I don't quite get the example in [Predicting on a new relation](http://psikit.net/spec/#predicting-on-a-new-relation). It might be an implementation detail, but in practice it will be quite hard to exchange data/models among supposedly interoperable services. I.e. in the POST example, how will example.org and flowers.com interact? Does example.org get the data or does flowers.com get the model? How does the _new attribute_ (flowers.com/data/irises/example.org/infer/knn_iris_20130802180742606) predict the values? Lazily? Can we have decisively separate services e.g. for hosting data relations (high storage, low CPU use) and for prediction (low storage, high CPU use)?
7. I find listed [future ideas](http://psikit.net/spec/#future-revisions) important.

In general, it seems this specification caters particularly to interactive API use with no provisions for workflow description (serialization, export), which is what would be needed to describe e.g. an OpenML workflow. There are already several competing specifications (PMML, [OntoDM](http://www.ontodm.com/), [ML-Schema](https://htmlpreview.github.io/?https://github.com/ML-Schema/documentation/blob/gh-pages/ML%20Schema.html) ([repo](https://github.com/ML-Schema/core)))  in this field, however.

**Comment by brylie** [comment]
@kernc thanks for your rather thorough review, and for the links to the other DM/ML schemas.

As a thought-experiment, here is a mock user interface example of how it might look to design machine learning flows, using granular components:

![screenshot_20170424_174117](https://cloud.githubusercontent.com/assets/17307/25478150/a9581a32-2b48-11e7-90db-aaf465e987b9.png)

The UI/programming environment in the above diagram is called [Node-Red](http://nodered.org), which is build on Node.js and can hook into system processes via standard in/out.

**Comment by kernc** [comment]
The library looks good and well supported. And it relies on the most common jQuery stuff, so that's a huge :+1: from me. My only remaining concern is whether it can be white-labelled enough to "sell" a Data Mining Suite built using Node Red and not vice versa. :smile:

**Comment by lopezco** [comment]
@kernc for workflow serialization/export there's also [PFA](http://dmg.org/pfa/docs/motivation/) format.
There is an issue about this #1789 with more information.

**Comment by brylie** [comment]
> My only remaining concern is whether it can be white-labelled enough to "sell" a Data Mining Suite built using Node Red and not vice versa.

From what I understand, there are a few whitelabel products that use Node-Red as a basis. I will do a bit of research on this idea.

In any case, there is clearly an opportunity in the Node-Red ecosystem to build/experiment with some scikit-learn, etc. integration.

**Comment by brylie** [comment]
After a brief search, I was able to find several domain-specific applications for Node-Red:

- ['cloud' data integration and dashboard](http://developers.sensetecnic.com/article/node-red-flow-using-freeboard/)
- [Industrial automation alerting](http://industrialinternet.co.uk/iot-2/node-red-knx-enocean/)
- [EU initiative for Smart City data integration and management](http://vital-iot.eu/blog/milestone-development)

 I have found only one white-labeled service, offered by IBM, called [Bluemix](https://dzone.com/articles/composing-apis-with-node-red-and-javascript).

**Comment by brylie** [comment]
It seems to me that integrating some data mining/machine learning tools in to these IoT data streams could provide a lot of value.

**Comment by john-bonachon** [comment]
Hey, a web version would be great.
I know there's Orange widgets documentation, but haven't found about Orange library itself.

It would be great to work on Orange GUI, save the workspace, and be able to load it in Python using the library (without GUI) and interact with it.
Then it would be possible to make a web service.

http request -> predict using Orange Python lib (using workspace loaded) -> http response

**Comment by brylie** [comment]
> It would be great to work on Orange GUI, save the workspace, and be able to load it in Python using the library (without GUI) and interact with it. Then it would be possible to make a web service.

This is a very interesting idea. Basically, Orange could be used as a visual machine learning service design tool.

**Comment by kernc** [comment]
Use case tracked as https://github.com/biolab/orange3/issues/1341. Some work done in https://github.com/biolab/orange3/pull/1449 but eventually abandoned. Would first need a stabler widget API if this should be maintained.

**Comment by brylie** [comment]
Below, is a mock-up of an hypothetical machine-learning UI that might be built with [WireCloud](https://github.com/Wirecloud/wirecloud):
![screenshot_20170529_183305](https://cloud.githubusercontent.com/assets/17307/26555463/d946158c-449d-11e7-9827-414c95dfd857.png)

![screenshot_20170529_183907](https://cloud.githubusercontent.com/assets/17307/26555570/4f1c9bc8-449e-11e7-8236-dffabca8c1f9.png)

WireCloud provides a drag-and-drop interface to build application 'mashups'. It is based on HTML/JavaScript, with a Django backend server. It can integrate with pretty much any REST service, and other types of services.

If we can build a functional Machine Learning (i.e. scikit-learn) REST API, we could likely create a visual design layer for building machine learning flows using WireCloud.

**Comment by brylie** [comment]
> It would be great to work on Orange GUI, save the workspace, and be able to load it in Python using the library (without GUI) and interact with it. Then it would be possible to make a web service.

In our previous experiment, we showed how Node-Red tool could be used to create a basic transit map service using real-time data:

[![apinf-experimentationmonday-nodered-2017-05-08](https://cloud.githubusercontent.com/assets/17307/26786523/9b096bd0-4a0f-11e7-98bd-c8c9e92e299d.jpg)](https://youtu.be/9P0JJ9mLZbg)

I didn't realize the previous experiment would be relevant to this discussion until we started forming the idea that we might create a **visual machine learning service design tool**.

Today's experiment consisted of starting a [design-first API specification based on the scikit-learn Python API](http://petstore.swagger.io/?url=https://raw.githubusercontent.com/apinf/ml-rest/master/doc/ml-rest.yaml). I welcome any constructive feedback regarding strengths and weaknesses of this design concept.

Please see the following video for a further explanation of the progress and plans for this experimentation series:

[![apinf-experimentationmonday-rest-and-ui-mockups-2017-06-05](https://cloud.githubusercontent.com/assets/17307/26786271/b8645c0e-4a0e-11e7-83e6-384b42c5ca08.jpg)](https://youtu.be/iEhDujPxubs)

**Comment by alitrack** [comment]
@brylie Is there any demo to try?

**Comment by navsh** [comment]
@ajdapretnar did you explore the idea of using Dash by Plot.ly? 
https://medium.com/@plotlygraphs/introducing-dash-5ecf7191b503 
https://github.com/plotly/dash

**Comment by brylie** [comment]
> did you explore the idea of using Dash by Plot.ly?

@navsh really great suggestion! Coming from a Meteor.js background, I am a fan of reactive programming, as well as declarative programming.

**Comment by brylie** [comment]
We are back from Summer holiday here in Finland. 

I made a bit more progress on the idea of **machine learning as a service with Python**. The following video shows a working example inspired by this discussion:

![[machine learning as a service with Python](https://youtu.be/MFD_T_9FyUc)](https://user-images.githubusercontent.com/17307/29362487-ed239e8c-8294-11e7-948c-b34965c71e8c.jpeg)

https://youtu.be/MFD_T_9FyUc

**Comment by praveenmaniyan** [comment]
@brylie have you come across clowdflows.org. It seems to be something that you are discussing about. You can also check the latest migration/implementation that is planned https://github.com/search?q=topic%3Aclowdflows+org%3Axflows&type=Repositories

**Comment by brylie** [comment]
Thanks @praveenmaniyan, I will check it out.

I have been reflecting a bit recently, and realize that visual programming is not ideally suited as a replacement for procedural style code. Some low level logic, such as flow control, may be useful to express with visual flow, but soon gets 'tangled' when working visually at that level of detail. So, the higher level of abstraction that Orange UI currently offers seems to be more productive.

What are your thoughts on visual programming environments that strike the 'right' level of abstraction?

**Comment by hmatejx** [comment]
Lately there seems to be development on the Qt side to bring Qt-based App on the web: http://blog.qt.io/blog/tag/qt-for-webassembly/

Furthermore, there is an alternative Wt library which is influenced by Qt, but allows for web-based rendering of the app: https://www.webtoolkit.eu/widgets

**Comment by janezd** [comment]
I like these ideas and they've been floating around in the core team for ten years already. After learning (the basics of) React, I'm tempted to try a small proof-of-principle myself.

I'm closing the issue, though. It is about a completely new project that is unlikely to be able to reuse any of existing code in Orange, and even if it does, the project would have its own repository with its own issues.

**Comment by bitsnaps** [comment]
@brylie 
> What are your thoughts on visual programming environments that strike the 'right' level of abstraction?
sorry not to re-open the issue, just want to point out [flask_jsondash](https://github.com/christabor/flask_jsondash) since it could an alternative to a starting point, I'm not sure if the workflow can be drawn using some Visual Programming block-based library (things like [blockly](https://github.com/google/blockly) or [Snap](https://github.com/jmoenig/Snap)).

**Comment by brylie** [comment]
@bitsnaps thanks for following up :-) I will check out those resources.

**Comment by brylie** [comment]
Just one off-hand thought, I think block-based and graph-based visual code editors are at the wrong level of abstraction when users are writing procedural code. Things might get very tangled or 'blocky'. Whereas, visual 'programming' seems more powerful when giving users high-level, almost declarative abstractions -- Orange seems to be designed at this level of abstraction.

**Comment by tdoan2010** [comment]
Hi, is there any progress on this? We're planning to have Orange as a Service, but it's not possible without a web-based version.

**Comment by ajdapretnar** [comment]
No, there's no progress on our side, but we are happy to support anyone who would tackle this task.

**Comment by tdoan2010** [comment]
That would be great. What is your preferred communication method? Email, chat channel, or something else?

**Comment by bitsnaps** [comment]
@brylie I think the closest concept to this idea can be [orchest](https://orchest.io/), it's flexible enough to orchestrate a data pipeline workflow, and yes, it is [open source](https://github.com/orchest/orchest) :) so you can try it online on [Gitpod](https://gitpod.io/#https://github.com/orchest/orchest).

**Comment by astrojuanlu** [comment]
Thanks for the mention @bitsnaps! (Juanlu from Orchest here). Notice that Orchest is [undergoing a port to Kubernetes](https://github.com/orchest/orchest/pull/623) (as explained [here](https://github.com/orchest/orchest/issues/737#issuecomment-1048215539)), but it's awesome that it currently works on Gitpod and we'll try to retain easy ways to deploy Orchest locally.

**Comment by nilsonfh** [comment]
What about [polyture](https://docs.polyture.com) ?

**Comment by bitsnaps** [comment]
> What about [polyture](https://docs.polyture.com) ?

It looks nice for ETL/Reporting, however it's not intended to be used specifically for Data Science ML pipeline projects, although it provides some algorithms (e.g. Sentiment Analysis...), and as I can see it's not fully open source..., you can try it on Gitpod [here](https://github.com/bitsnaps/polyture.github.io/tree/gitpod-setup).
